node/src/components/linear_chain.rs

```diff
@@ -197,28 +269,61 @@ where
                 let era_id = block_header.era_id();
                 let height = block_header.height();
                 info!(?block_hash, ?era_id, ?height, "Linear chain block stored.");
-                let mut effects = effect_builder.put_execution_results_to_storage(block_hash, execution_results).ignore();
+                let mut effects = effect_builder
+                    .put_execution_results_to_storage(block_hash, execution_results)
+                    .ignore();
                 effects.extend(
-                    effect_builder.handle_linear_chain_block(block_header.clone())
-                    .event(move |signature| Event::NewFinalitySignature(block_hash, signature)));
-                effects.extend(effect_builder.announce_block_added(block_hash, block_header).ignore());
+                    effect_builder
+                        .handle_linear_chain_block(block_header.clone())
+                        .event(move |fs| Event::NewFinalitySignature(Box::new(fs))),
+                );
+                effects.extend(
+                    effect_builder
+                        .announce_block_added(block_hash, block_header)
+                        .ignore(),
+                );
                 effects
-            },
-            Event::NewFinalitySignature(block_hash, signature) => {
+            }
+            Event::NewFinalitySignature(fs) => {
+                let FinalitySignature {
+                    block_hash,
+                    signature,
+                    public_key,
+                } = *fs;
+                let verified =
+                    crypto::asymmetric_key::verify(block_hash.inner(), &signature, &public_key);
```

Comment:
That would also be deduplicated if it was just `fs.verify()`. (Or should we even verify on deserialization?)

Reply:
I don't control the deserialization process, `FinalitySignature` is on of the `protocol::Message` variants. I could verify in the reactor but I am not sure if that's the place to do it. I can add a `verify()` method on the `FinalitySignature` itself and call it here.

Reply:
Done in [`0e1f67a` (#577)](https://github.com/CasperLabs/casper-node/pull/577/commits/0e1f67acc1a4c05f99d775e578dfc02fc636a115)